x-pack/plugins/lens/public/indexpattern_datasource/utils.tsx

```diff
@@ -238,3 +242,127 @@ export function getVisualDefaultsForLayer(layer: IndexPatternLayer) {
     {}
   );
 }
+
+export function getFiltersInLayer(
+  layer: IndexPatternLayer,
+  columnIds: string[],
+  layerData: NonNullable<FramePublicAPI['activeData']>[string] | undefined
+) {
+  // extract filters from filtered metrics
+  // consider all the columns, included referenced ones to cover also the formula case
+  const filteredMetrics = Object.keys(layer.columns)
+    .map((colId) => {
+      // there's a special case to handle when a formula has a global filter applied
+      // in this case ignore the filter on the formula column and only use the filter
+      // applied to the referenced columns.
+      // This will avoid duplicate filters and issues when a global formula filter is
+      // combine to specific referenced columns filters
+      if (
+        isColumnOfType<FormulaIndexPatternColumn>('formula', layer.columns[colId]) &&
+        layer.columns[colId]?.filter
+      ) {
+        return null;
+      }
+      return layer.columns[colId]?.filter;
+    })
+    // filter out empty filters as well
+    .filter((filter) => filter?.query?.trim()) as Query[];
+
+  const { kuery: kqlMetricQueries, lucene: luceneMetricQueries } = groupBy(
+    filteredMetrics,
+    'language'
+  );
+
+  function extractUsefulQueries(
+    queries: FiltersIndexPatternColumn['params']['filters'] | undefined
+  ) {
+    return queries?.map(({ input }) => input).filter(({ query }) => query?.trim() && query !== '*');
+  }
+
+  const filterOperation = columnIds
+    .map((colId) => {
+      const column = layer.columns[colId];
+
+      if (isColumnOfType<FiltersIndexPatternColumn>('filters', column)) {
```

Comment:
nit: I think it would be clearer that this is a set of mutually-exclusive cases if the `if` blocks hung together with `else if`'s and there was just a single return statement at the end.

Reply:
I'm not a big fan of either single return point for complex cases, nor of the `else if` branching if not driven by specific performance issues.
But I take the point on readability here: I've refactored the code to be more encapsulated with isolated functions per column type. Hope now to have made it more clear to read/understand.